In [4]:
import pandas as pd
import numpy as np

In [5]:
df = pd.read_csv("netflix_titles.csv")

In [6]:
print(df.head())
print(df.info())
print(df.describe())
print(df.shape)

  show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021        

In [7]:
print(df.dtypes)

show_id           str
type              str
title             str
director          str
cast              str
country           str
date_added        str
release_year    int64
rating            str
duration          str
listed_in         str
description       str
dtype: object


In [8]:
print(df.isnull().sum())

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64


In [9]:
missing = df.isnull().sum() / len(df) * 100
print(missing.sort_values(ascending=False))

director        29.908028
country          9.435676
cast             9.367549
date_added       0.113546
rating           0.045418
duration         0.034064
show_id          0.000000
type             0.000000
title            0.000000
release_year     0.000000
listed_in        0.000000
description      0.000000
dtype: float64


In [10]:
print(df.duplicated().sum())

df.drop_duplicates(inplace=True)

0


In [11]:
df["director"] = df["director"].fillna("Unknown")

In [12]:
df["country"] = df["country"].fillna("Unknown")

In [13]:
df.dropna(subset=["date_added"], inplace=True)

In [14]:
print(df["date_added"].dtype)

str


In [15]:
df["date_added"] = pd.to_datetime(df["date_added"], errors="coerce")

In [16]:
df["date_added"].isna().sum()

np.int64(88)

In [17]:
df[df["date_added"].isna()]

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
6079,s6080,TV Show,Abnormal Summit,"Jung-ah Im, Seung-uk Jo","Hyun-moo Jun, Si-kyung Sung, Se-yoon Yoo",South Korea,NaT,2017,TV-PG,2 Seasons,"International TV Shows, Korean TV Shows, Stand...","Led by a trio of Korean celebs, a multinationa..."
6177,s6178,TV Show,忍者ハットリくん,Unknown,NaN,Japan,NaT,2012,TV-Y7,2 Seasons,"Anime Series, Kids' TV","Hailing from the mountains of Iga, Kanzo Hatto..."
6213,s6214,TV Show,Bad Education,Unknown,"Jack Whitehall, Mathew Horne, Sarah Solemani, ...",United Kingdom,NaT,2014,TV-MA,3 Seasons,"British TV Shows, TV Comedies","A history teacher at the posh Abbey Grove, Alf..."
6279,s6280,TV Show,Being Mary Jane: The Series,Unknown,"Gabrielle Union, Lisa Vidal, Margaret Avery, O...",United States,NaT,2016,TV-14,4 Seasons,"Romantic TV Shows, TV Dramas",Ambitious single TV journalist Mary Jane attem...
6304,s6305,TV Show,"Big Dreams, Small Spaces",Unknown,Monty Don,United Kingdom,NaT,2017,TV-G,3 Seasons,"British TV Shows, International TV Shows, Real...",Writer and presenter Monty Don helps England's...
...,...,...,...,...,...,...,...,...,...,...,...,...
8539,s8540,TV Show,The Tudors,Unknown,"Jonathan Rhys Meyers, Henry Cavill, James Frai...","Ireland, Canada, United States, United Kingdom",NaT,2010,TV-MA,4 Seasons,TV Dramas,All the splendor and scandal of England's 16th...
8557,s8558,TV Show,The West Wing,Unknown,"Martin Sheen, Rob Lowe, Allison Janney, John S...",United States,NaT,2005,TV-14,7 Seasons,TV Dramas,This powerful political epic chronicles the tr...
8684,s8685,TV Show,Vroomiz,Unknown,"Joon-seok Song, Jeong-hwa Yang, Sang-hyun Um, ...",South Korea,NaT,2016,TV-Y,3 Seasons,"Kids' TV, Korean TV Shows","For these half-car, half-animal friends, each ..."
8712,s8713,TV Show,Weird Wonders of the World,Unknown,Chris Packham,United Kingdom,NaT,2016,TV-PG,2 Seasons,"British TV Shows, Docuseries, Science & Nature TV",From animal oddities and bizarre science to me...


In [18]:
df["date_added"] = pd.to_datetime(
    df["date_added"].astype(str).str.strip(),
    errors="coerce"
)

In [19]:
print(df["date_added"].dtype)

datetime64[us]


In [20]:
df["date_added"].isna().sum() / len(df) * 100


np.float64(1.000341025349551)

In [21]:
missing_count = df["date_added"].isna().sum()
missing_percent = missing_count / len(df) * 100

print("Brakujące daty:", missing_count)
print("Procent braków:", round(missing_percent, 2), "%")

Brakujące daty: 88
Procent braków: 1.0 %


In [22]:
df[df["date_added"].isna()]

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
6079,s6080,TV Show,Abnormal Summit,"Jung-ah Im, Seung-uk Jo","Hyun-moo Jun, Si-kyung Sung, Se-yoon Yoo",South Korea,NaT,2017,TV-PG,2 Seasons,"International TV Shows, Korean TV Shows, Stand...","Led by a trio of Korean celebs, a multinationa..."
6177,s6178,TV Show,忍者ハットリくん,Unknown,NaN,Japan,NaT,2012,TV-Y7,2 Seasons,"Anime Series, Kids' TV","Hailing from the mountains of Iga, Kanzo Hatto..."
6213,s6214,TV Show,Bad Education,Unknown,"Jack Whitehall, Mathew Horne, Sarah Solemani, ...",United Kingdom,NaT,2014,TV-MA,3 Seasons,"British TV Shows, TV Comedies","A history teacher at the posh Abbey Grove, Alf..."
6279,s6280,TV Show,Being Mary Jane: The Series,Unknown,"Gabrielle Union, Lisa Vidal, Margaret Avery, O...",United States,NaT,2016,TV-14,4 Seasons,"Romantic TV Shows, TV Dramas",Ambitious single TV journalist Mary Jane attem...
6304,s6305,TV Show,"Big Dreams, Small Spaces",Unknown,Monty Don,United Kingdom,NaT,2017,TV-G,3 Seasons,"British TV Shows, International TV Shows, Real...",Writer and presenter Monty Don helps England's...
...,...,...,...,...,...,...,...,...,...,...,...,...
8539,s8540,TV Show,The Tudors,Unknown,"Jonathan Rhys Meyers, Henry Cavill, James Frai...","Ireland, Canada, United States, United Kingdom",NaT,2010,TV-MA,4 Seasons,TV Dramas,All the splendor and scandal of England's 16th...
8557,s8558,TV Show,The West Wing,Unknown,"Martin Sheen, Rob Lowe, Allison Janney, John S...",United States,NaT,2005,TV-14,7 Seasons,TV Dramas,This powerful political epic chronicles the tr...
8684,s8685,TV Show,Vroomiz,Unknown,"Joon-seok Song, Jeong-hwa Yang, Sang-hyun Um, ...",South Korea,NaT,2016,TV-Y,3 Seasons,"Kids' TV, Korean TV Shows","For these half-car, half-animal friends, each ..."
8712,s8713,TV Show,Weird Wonders of the World,Unknown,Chris Packham,United Kingdom,NaT,2016,TV-PG,2 Seasons,"British TV Shows, Docuseries, Science & Nature TV",From animal oddities and bizarre science to me...


In [23]:
df.loc[df["date_added"].isna(), "date_added"]

6079   NaT
6177   NaT
6213   NaT
6279   NaT
6304   NaT
        ..
8539   NaT
8557   NaT
8684   NaT
8712   NaT
8755   NaT
Name: date_added, Length: 88, dtype: datetime64[us]

In [24]:
df["date_added_raw"] = df["date_added"]

df["date_added"] = pd.to_datetime(df["date_added"], errors="coerce")

df[df["date_added"].isna()][["date_added_raw"]]

,date_added_raw
6079,NaT
6177,NaT
6213,NaT
6279,NaT
6304,NaT
...,...
8539,NaT
8557,NaT
8684,NaT
8712,NaT


In [25]:
print("Analiza date_added zakończona – brakujące wartości pozostawiono bez imputacji")


Analiza date_added zakończona – brakujące wartości pozostawiono bez imputacji


Analiza i czyszczenie kolumny 'cast'

In [26]:
print("Brakujące wartości w kolumnie cast:")
print(df["cast"].isnull().sum())

Brakujące wartości w kolumnie cast:
825


In [27]:
df.loc[df["cast"].isna(), ["title", "director", "cast"]]

,title,director,cast
0,Dick Johnson Is Dead,Kirsten Johnson,NaN
3,Jailbirds New Orleans,Unknown,NaN
10,"Vendetta: Truth, Lies and The Mafia",Unknown,NaN
14,Crime Stories: India Detectives,Unknown,NaN
16,Europe's Most Dangerous Man: Otto Skorzeny in ...,"Pedro de Echave García, Pablo Azorín Williams",NaN
...,...,...,...
8746,Winnie,Pascale Lamche,NaN
8755,Women Behind Bars,Unknown,NaN
8756,Woodstock,Barak Goodman,NaN
8763,WWII: Report from the Aleutians,John Huston,NaN


In [28]:
df["cast"] = df["cast"].fillna("Unknown")

In [29]:
print(df["cast"].isnull().sum())

0


In [30]:
print("Brakujące wartości w kolumnie cast:", df["cast"].isnull().sum())

Brakujące wartości w kolumnie cast: 0


### Wnioski

- W kolumnie `cast` znajdowało się 825 brakujących wartości.
- Braki zostały uzupełnione wartością `"Unknown"`.
- Po czyszczeniu liczba brakujących wartości wynosi 0.

## Analiza i czyszczenie kolumny `rating`

In [31]:
print("Brakujące wartości w kolumnie rating:")
print(df["rating"].isnull().sum())


Brakujące wartości w kolumnie rating:
4


In [32]:
df.loc[df["rating"].isna(), ["title", "rating"]]

,title,rating
5989,13TH: A Conversation with Oprah Winfrey & Ava ...,NaN
6827,Gargantia on the Verdurous Planet,NaN
7312,Little Lunch,NaN
7537,My Honor Was Loyalty,NaN


In [33]:
df["rating"] = df["rating"].fillna("Unknown")

In [34]:
print(df["rating"].isnull().sum())

0


Sprawdzenie końcowe

In [36]:
print(df.isnull().sum())

show_id            0
type               0
title              0
director           0
cast               0
country            0
date_added        88
release_year       0
rating             0
duration           3
listed_in          0
description        0
date_added_raw    88
dtype: int64


In [37]:
df["duration"] = df["duration"].fillna("Unknown")

In [38]:
print(df["duration"].isnull().sum())

0


In [39]:
print(df.isnull().sum())

show_id            0
type               0
title              0
director           0
cast               0
country            0
date_added        88
release_year       0
rating             0
duration           0
listed_in          0
description        0
date_added_raw    88
dtype: int64
